lib/std/cli.zig

```diff
@@ -0,0 +1,1157 @@
+                                        return usageError(writer, "unable to parse --{s}={s}: {s}", .{ field.name, arg_value, @errorName(err) });
+                                    });
+                                },
+                                .int => {
+                                    try array_list.append(allocator, std.fmt.parseInt(ptrInfo.child, arg_value, 0) catch |err| {
+                                        return usageError(writer, "unable to parse --{s}={s}: {s}", .{ field.name, arg_value, @errorName(err) });
+                                    });
+                                },
+                                .@"enum" => comptime unreachable,
+                                .pointer => |ptrInfo2| {
+                                    comptime assert(ptrInfo2.size == .slice);
+                                    if (ptrInfo2.child == u8) {
+                                        // String.
+                                        try array_list.append(allocator, arg_value); // To resolve compile errors between `[:0]const u8` and `[]const u8` on this line, ensure the passed-in args are `[:0]const u8`.
+                                    } else comptime unreachable;
+                                },
+                                else => comptime unreachable,
+                            }
+                        }
+                    },
+                    else => comptime unreachable,
+                }
+                fields_seen[i] = true;
+                break;
+            }
+        } else {
+            // Didn't match anything.
+            return usageError(writer, "unrecognized argument: {s}", .{arg});
+        }
+    }
+
+    // Fill default values.
+    inline for (named_info.fields, 0..) |field, i| {
+        if (!fields_seen[i]) {
+            if (field.defaultValue()) |default| {
+                @field(result.named, field.name) = default;
+            } else {
+                if (field.type == bool) {
+                    return usageError(writer, "missing required argument: --" ++ field.name ++ " or --no-" ++ field.name, .{});
+                } else {
+                    return usageError(writer, "missing required argument: --" ++ field.name, .{});
+                }
+            }
+        }
+    }
+
+    // Finalize the array lists.
+    result.positional = try positional.toOwnedSlice(allocator);
+    inline for (@typeInfo(@TypeOf(array_lists)).@"struct".fields) |field| {
+        @field(result.named, field.name) = try @field(array_lists, field.name).toOwnedSlice(allocator);
+    }
+
+    return result;
+}
+
+fn checkArgsType(comptime Args: type) void {
+    const args_fields = @typeInfo(Args).@"struct".fields;
+    if (!(args_fields.len == 2 and mem.eql(u8, args_fields[0].name, "named") and mem.eql(u8, args_fields[1].name, "positional"))) @compileError("expected Args to have exactly these fields in this order: named, positional");
```

Comment:
requiring the order here seems questionable maybe.

Reply:
It is. People WILL stumble upon it since they (understandably) think of structs more similar to a hashmap than an array.

But using `std.sort.block` and `std.sort.binarySearch` (there is no other search function in the standard library where you can define the predicate afaik) could help here.